### PR TITLE
Replace variables in delegate_to template. Fixes #20508

### LINKED ISSUE #20508: Using set_fact with delegate_to, delegate_facts and with_items only sets the fact on the first host
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

set_fact

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = 
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

Red Hat Enterprise Linux Client release 7.3 (Maipo)

##### SUMMARY

In Ansible 2.2.1.0, if you use set_fact, delegate_to and delegate_facts to set facts on several hosts listed in with_items, only the first host in the list will have the fact set.  In all other versions of Ansible 2.x, the fact will be set on all hosts.

##### STEPS TO REPRODUCE

Create a playbook like this:
```yaml
---

- hosts: localhost
  tasks:
  - name: "Set fact on localhost"
    set_fact:
      my_test_fact: "Hello"

  - name: "Copy fact to other servers"
    set_fact:
      my_test_fact: "{{ my_test_fact }}"
    delegate_to: "{{ item }}"
    delegate_facts: True
    with_items: "{{ groups['all'] }}"
    when: item != "localhost"

- hosts: server1
  gather_facts: True
  tasks:
  - name: "Show fact on server1"
    debug: var=my_test_fact

- hosts: server2
  gather_facts: True
  tasks:
  - name: "Show fact on server2"
    debug: var=my_test_fact
```
I am using a simple inventory file:
```
[server1]
192.168.1.105 ansible_user=admin

[server2]
192.168.1.106 ansible_user=admin
```

##### EXPECTED RESULTS

If I run the playbook with Ansible 2.2.0.0 I get these results:
```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Set fact on localhost] ***************************************************
ok: [localhost]

TASK [Copy fact to other servers] **********************************************
ok: [localhost -> 192.168.1.105] => (item=192.168.1.105)
ok: [localhost -> 192.168.1.106] => (item=192.168.1.106)

PLAY [server1] *****************************************************************

TASK [setup] *******************************************************************
ok: [192.168.1.105]

TASK [Show fact on server1] ****************************************************
ok: [192.168.1.105] => {
    "my_test_fact": "Hello"
}

PLAY [server2] *****************************************************************

TASK [setup] *******************************************************************
ok: [192.168.1.106]

TASK [Show fact on server2] ****************************************************
ok: [192.168.1.106] => {
    "my_test_fact": "Hello"
}

PLAY RECAP *********************************************************************
192.168.1.105              : ok=2    changed=0    unreachable=0    failed=0   
192.168.1.106              : ok=2    changed=0    unreachable=0    failed=0   
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

##### ACTUAL RESULTS
If I run with Ansible 2.2.1.0, the fact is only set for the first host:
```

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Set fact on localhost] ***************************************************
ok: [localhost]

TASK [Copy fact to other servers] **********************************************
ok: [localhost -> 192.168.1.105] => (item=192.168.1.105)
ok: [localhost -> 192.168.1.106] => (item=192.168.1.106)

PLAY [server1] *****************************************************************

TASK [setup] *******************************************************************
ok: [192.168.1.105]

TASK [Show fact on server1] ****************************************************
ok: [192.168.1.105] => {
    "my_test_fact": "Hello"
}

PLAY [server2] *****************************************************************

TASK [setup] *******************************************************************
ok: [192.168.1.106]

TASK [Show fact on server2] ****************************************************
ok: [192.168.1.106] => {
    "my_test_fact": "VARIABLE IS NOT DEFINED!"
}

PLAY RECAP *********************************************************************
192.168.1.105              : ok=2    changed=0    unreachable=0    failed=0   
192.168.1.106              : ok=2    changed=0    unreachable=0    failed=0   
localhost                  : ok=3    changed=0    unreachable=0    failed=0   

```

The bug seems to be related to a change to plugins/strategy/__init__.py here: https://github.com/ansible/ansible/commit/2d8ebbfe8cc9332b34ce807b1c30e69507ba36a0

### PR DESCRIPTION
**Pull request by chrulri**
Logic based on Ansible 2.2.0.0 source, code adapted to changes introduced with #22003

##### SUMMARY
Fixes #20508

Up to Ansible 2.2.0.0 the delegate_to was templated, e.g. use of with_items and delegate_to: "{{ item }}" was possible.
Ansible 2.2.1.0 and newer lack this functionality. {{ item }} does not get replaced and the facts gathered in case of "setup" module are stored for literal host "{{ item }}" instead of the actual delegated host.
This change is based upon the changes in #22003 and adds the missing functionality of Ansible 2.2.0.0 into `get_delegated_hosts`.

##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
strategy plugin

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### ADDITIONAL INFORMATION
See #20508 for more details.

###### Playbook
```
- hosts: localhost
  gather_facts: false
  tasks:
    - name: "gather data from nodes"
      setup:
      delegate_to: "{{ item }}"
      delegate_facts: true
      with_items: "{{ groups.all }}"

    - name: "print ipv4s"
      debug: var=hostvars[item].ansible_all_ipv4_addresses
      with_items: "{{ groups.all }}"
```

###### Inventory
```
node1      ansible_ssh_host=172.17.2.11
node2      ansible_ssh_host=172.17.2.12
```

###### Before (w/o patch)
```
$ ansible-playbook -i inventory.yml simple.yml

PLAY [localhost] ******************************************************************************************************************************************************************

TASK [gather data from nodes] *****************************************************************************************************************************************************
ok: [localhost -> None] => (item=node1)
ok: [localhost -> None] => (item=node2)

TASK [print ipv4s] ****************************************************************************************************************************************************************
ok: [localhost] => (item=node1) => {
    "hostvars[item].ansible_all_ipv4_addresses": "VARIABLE IS NOT DEFINED!",
    "item": "node1"
}
ok: [localhost] => (item=node2) => {
    "hostvars[item].ansible_all_ipv4_addresses": "VARIABLE IS NOT DEFINED!",
    "item": "node2"
}

PLAY RECAP ************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

###### After (w/ patch)
```
$ ansible-playbook -i inventory.yml simple.yml

PLAY [localhost] ******************************************************************************************************************************************************************

TASK [gather data from nodes] *****************************************************************************************************************************************************
ok: [localhost -> None] => (item=node1)
ok: [localhost -> None] => (item=node2)

TASK [print ipv4s] ****************************************************************************************************************************************************************
ok: [localhost] => (item=node1) => {
    "hostvars[item].ansible_all_ipv4_addresses": [
        "172.17.2.11"
    ],
    "item": "node1"
}
ok: [localhost] => (item=node2) => {
    "hostvars[item].ansible_all_ipv4_addresses": [
        "172.17.2.12"
    ],
    "item": "node2"
}

PLAY RECAP ************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

### CODE REVIEW

**Comment by sosiouxme** [comment]
This patch appears to fix my similar issue, which oddly enough seems to have been introduced in 2.2.2.0 (from pip install) not 2.2.1.0 but in any case manifests the same (delegated facts not getting set `with_items`). +1 to getting this in.

**Comment by chrulri** [comment]
@sosiouxme As far as I understood the commit history:
- The templar line was removed in commit [2d8ebbf](https://github.com/ansible/ansible/commit/2d8ebbfe8cc9332b34ce807b1c30e69507ba36a0#commitcomment-21028761) which was first included in 2.2.1.0.
- In 2.3.0.0 there was some reworking on the code (see b3abab1), but the delegate_to was never templated again.

This PR here is for devel branch, but one could backport this fix to 2.2 if necessary. All the functionality is still there, just the code that hold it all together was removed for a reason I don't know.

**Comment by Jorge-Rodriguez** [comment]
This PR seems to be very similar to the change that fixed #17582 and was later removed.

**Comment by bcoca** [comment]
@mattclay not sure how to test this, but we should include to avoid breaking it again

**Comment by jimi-c** [comment]
Hi @chrulri, I decided to fix this in a slightly different way, which ended up being quite a bit more simple (and should be much quicker too, as templating things can introduce slowness to results processing).

Thanks for the PR, which definitely made it much easier to pinpoint the problem!

**Comment by Jorge-Rodriguez** [comment]
@jimi-c Which release will this be on?

**Comment by jordiclariana** [comment]
Hi, I've been playing a little with this fix and I detected this problem:

Ansible version used:
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'modules']
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```
test.yml:
```
- hosts: localhost
  connection: local
  tasks:
    - include: "{{loop1}}"
      with_items:
        - task1.yml
        - task2.yml
      loop_control:
        loop_var: loop1
```

task1.yml:
```
- debug: msg="Task1"
```

task2.yml:
```
- debug: msg="Task2"
```

Run command: `ansible-playbook -ilocalhost, test.yml -vvv`
```
Using /etc/ansible/ansible.cfg as config file

PLAYBOOK: test.yml **********************************************************************************************************************************************************
1 plays in test.yml

PLAY [localhost] ************************************************************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************************************************************
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: jordi
<localhost> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
ok: [localhost]
META: ran handlers

TASK [include] **************************************************************************************************************************************************************
task path: /tmp/ansible_test2/test.yml:6
ERROR! Unexpected Exception: 'dict' object has no attribute 'loop_var'
the full traceback was:

Traceback (most recent call last):
  File "/usr/local/bin/ansible-playbook", line 109, in <module>
    exit_code = cli.run()
  File "/usr/local/lib/python2.7/dist-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/playbook_executor.py", line 153, in run
    result = self._tqm.run(play=play)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_queue_manager.py", line 284, in run
    play_return = strategy.run(iterator, play_context)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/strategy/linear.py", line 317, in run
    variable_manager=self._variable_manager
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/included_file.py", line 86, in process_include_results
    loop_var = original_task.loop_control.loop_var or 'item'
AttributeError: 'dict' object has no attribute 'loop_var'
```

These very same playbook & tasks work without the patch. I also noticed (so far) that it only fails with an `include` using `with_items` and `loop_control`.

**Comment by chrulri** [comment]
@jordiclariana with "this fix" are you talking about this pullrequest #23599 which was never merged into any version of Ansible or are you talking about the change introduced by @jimi-c in version 2.3.1 [here](https://github.com/ansible/ansible/commit/e5cd675b38d49b22f1ecc929943d016fb7159856) ?

**Comment by jordiclariana** [comment]
The PR. I just downloaded the raw file from the [PR](https://raw.githubusercontent.com/chrulri/ansible/0de1bc7d6ba051a286a597eb17d76ac94a58ff1f/lib/ansible/plugins/strategy/__init__.py).

After this test I realized that the PR is way behind v2.3.1.0, so some adjustments must be done before a merge. I played a little bit with the PR and the original file from v2.3.1.0 and these are the changes that makes it work properly:

```
240,245c240,243
<     def get_delegated_hosts(self, result, task):
<         host_name = result.get('_ansible_delegated_vars', {}).get('ansible_host', None)
<         if host_name is not None:
<             actual_host = self._inventory.get_host(host_name)
<             if actual_host is None:
<                 actual_host = Host(name=host_name)
---
>     def get_delegated_hosts(self, result, original_host, play, task):
>         # set correct loop var
>         if task.loop_control:
>             loop_var = task.loop_control.loop_var or 'item'
247c245,257
<             actual_host = Host(name=task.delegate_to)
---
>             loop_var = 'item'
>         item = result.get(loop_var, None)
> 
>         task_vars = self._variable_manager.get_vars(loader=self._loader, play=play, host=original_host, task=task)
>         self.add_tqm_variables(task_vars, play=play)
>         if item is not None:
>             task_vars[loop_var] = item
> 
>         templar = Templar(loader=self._loader, variables=task_vars)
>         host_name = templar.template(task.delegate_to)
>         actual_host = self._inventory.get_host(host_name)
>         if actual_host is None:
>             actual_host = Host(name=host_name)
493c503
<                             host_list = self.get_delegated_hosts(result_item, original_task)
---
>                             host_list = self.get_delegated_hosts(result_item, original_host, iterator._play, original_task)
```

In a nutshell: forget my previous comment. What is important is that the PR (which fixes the original issue) works after some changes to adjust it to latest ansible stable version.
